src/rdf4cpp/rdf/storage/node/reference_node_storage/detail/BiDirFlatMap.hpp

```diff
@@ -0,0 +1,188 @@
+
+        [[no_unique_address]] Hash hash;
+
+        size_t operator()(backward_key_type const &a) const noexcept {
+            return a.hash;
+        }
+
+        size_t operator()(view_type const &a) const noexcept {
+            return hash(a);
+        }
+    };
+
+    using backward_allocator_type = typename std::allocator_traits<Allocator>::template rebind_alloc<backward_key_type>;
+    using backward_type = dice::sparse_map::sparse_set<backward_key_type, backward_hasher, backward_key_equal, backward_allocator_type>;
+
+    forward_type forward_;   // forward_[id] stores Value for Id id
+    backward_type backward_; // backward_[view] stores Id into forward_
+    [[no_unique_address]] allocator_type alloc_;
+
+    [[nodiscard]] static constexpr size_type to_index(id_type id) noexcept {
+        return static_cast<size_type>(id) - 1;
+    }
+
+    [[nodiscard]] static constexpr id_type to_id(size_type ix) noexcept {
+        return static_cast<id_type>(ix + 1);
+    }
+
+public:
+    explicit BiDirFlatMap(hasher const &hash = hasher{},
+                          key_equal const &equal = key_equal{},
+                          allocator_type const &alloc = allocator_type{}) noexcept : forward_{alloc},
+                                                                                     backward_{0, backward_hasher{hash}, backward_key_equal{&forward_, equal}, alloc},
+                                                                                     alloc_{alloc} {
+    }
+
+    explicit BiDirFlatMap(allocator_type const &alloc) noexcept : BiDirFlatMap{hasher{}, key_equal{}, alloc} {
+    }
+
+    // deleted because type is self-referential
+    BiDirFlatMap(BiDirFlatMap const &) = delete;
+    BiDirFlatMap(BiDirFlatMap &&) = delete;
+    BiDirFlatMap &operator=(BiDirFlatMap const &) = delete;
+    BiDirFlatMap &operator=(BiDirFlatMap &&) = delete;
+
+    [[nodiscard]] size_type size() const noexcept {
+        return forward_.size();
+    }
+
+    void shrink_to_fit() {
+        forward_.shrink_to_fit();
+        // TODO backward_?
+    }
+
+    [[nodiscard]] std::optional<view_type> lookup_value(id_type const id) const noexcept {
+        if (id == id_type{}) [[unlikely]] {
+            return std::nullopt;
+        }
+
+        auto const ix = to_index(id);
+        if (ix >= forward_.size() || !forward_[ix].has_value()) {
+            return std::nullopt;
+        }
+
+        return static_cast<view_type>(*forward_[ix]);
+    }
+
+    [[nodiscard]] id_type lookup_id(view_type const &view) const noexcept {
+        auto it = backward_.find(view);
+        if (it == backward_.end()) {
+            return Id{};
+        }
+
+        return it->id;
+    }
+
+    void reserve(id_type const min_id) {
+        forward_.resize(to_index(min_id));
+    }
+
+    [[nodiscard]] id_type insert_assume_not_present(view_type const &view) {
+        assert(lookup_id(view) == Id{});
+
+        auto const assigned_id = to_id(forward_.size());
```

Comment:
We had almost the same issue before (i.e. ids not being reused) just that it didn't occupy additional space. Didn't really think about that.

So I see two options:
1. keep the increasing id behaviour as is, apply an offset to the vector index so that we do not need to store the empty range at the start
2. Solve both problems by reusing ids

Reply:
I would reuse IDs. Let's have a short call about how to solve this tomorrow.

Reply:
Implemented